osu.Game.Rulesets.Osu/Objects/Drawables/DrawableOsuHitObject.cs

```diff
@@ -60,6 +61,13 @@ protected override void OnApply(HitObject hitObject)
             PositionBindable.BindTo(HitObject.PositionBindable);
             StackHeightBindable.BindTo(HitObject.StackHeightBindable);
             ScaleBindable.BindTo(HitObject.ScaleBindable);
+
+            // Manually set to reduce the number of future alive objects to a bare minimum.
+            LifetimeStart = HitObject.StartTime - HitObject.TimePreempt;
+
+            // Arbitrary lifetime end to prevent past objects in idle states remaining alive in non-frame-stable contexts.
+            // An extra 1000ms is added to always overestimate the true lifetime, and a more exact value is set by hit transforms and the following expiry.
+            LifetimeEnd = HitObject.GetEndTime() + HitObject.HitWindows.WindowFor(HitResult.Miss) + 1000;
```

Comment:
we already do very similar to this in scrolling containers right?

Reply:
Nah, scrolling containers also rely on the hitobjects to expire, because it is/used to be hard for them to tell when the appropriate expiry time should be.
They’ll need something similar though.